process/asyncExecution/queue/blocksQueue.go

```diff
@@ -0,0 +1,149 @@
+package queue
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/multiversx/mx-chain-core-go/core/check"
+	"github.com/multiversx/mx-chain-core-go/data"
+	"github.com/multiversx/mx-chain-go/common"
+	logger "github.com/multiversx/mx-chain-logger-go"
+)
+
+var log = logger.GetOrCreate("process/asyncExecution/queue")
+
+// blocksQueue represents a thread-safe queue for storing and processing blockchain headers.
+// It provides methods to add headers at the end or beginning of the queue and retrieve them
+// for processing in a FIFO (First In, First Out) manner
+type blocksQueue struct {
+	mutex           *sync.Mutex
+	headerBodyPairs []HeaderBodyPair
+	closed          bool
+	notifyCh        chan struct{} // used only for blocking
+}
+
+// NewBlocksQueue creates and returns a new instance of blocksQueue
+func NewBlocksQueue() (*blocksQueue, error) {
+	mutex := &sync.Mutex{}
+
+	return &blocksQueue{
+		mutex:           mutex,
+		headerBodyPairs: make([]HeaderBodyPair, 0),
+		notifyCh:        make(chan struct{}, 1), // buffered so send won't block if not read yet
+	}, nil
+}
+
+// AddOrReplace appends a HeaderBodyPair to the end of the queue,
+// or replaces the last element if it has the same nonce.
+func (bq *blocksQueue) AddOrReplace(pair HeaderBodyPair) error {
```

Comment:
I think we will need also a Clean exported function, that would for example clear the queue and set the "last popped block" to a certain given nonce (so that after the clean the next added blocks in the queue can be validated with the new context)

Reply:
done